src/TreeSitter/Importing.hs

```diff
@@ -1,135 +1,267 @@
-module TreeSitter.Importing where
+{-# LANGUAGE DefaultSignatures, FlexibleContexts, FlexibleInstances,
+             PolyKinds, ScopedTypeVariables, TypeApplications, TypeOperators #-}
+module TreeSitter.Importing
+( parseByteString
+, FieldName(..)
+, Unmarshal(..)
+, SymbolMatching(..)
+, step
+, push
+, goto
+, peekNode
+, peekFieldName
+, getFields
+) where
 
-import Control.Exception as Exc
-import Data.ByteString
-
-import           Data.ByteString.Unsafe (unsafeUseAsCStringLen)
+import           Control.Applicative
+import           Control.Effect hiding ((:+:))
+import           Control.Effect.Reader
+import           Control.Effect.Fail
+import           Control.Monad (void)
+import           Control.Monad.IO.Class
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import qualified Data.Map as Map
+import qualified Data.Text as Text
+import           Data.Text.Encoding
+import           Foreign.C.String
 import           Foreign.Marshal.Alloc
+import           Foreign.Marshal.Utils
 import           Foreign.Ptr
 import           Foreign.Storable
-import TreeSitter.Cursor as TS
-import TreeSitter.Node as TS
-import TreeSitter.Parser as TS
-import TreeSitter.Tree as TS
-import qualified Data.Text as Text
-import Control.Effect
-import Control.Effect.Reader
-import Control.Monad.IO.Class
-import Data.Text.Encoding
-import qualified Data.ByteString as B
-import Control.Applicative
-import Control.Monad (void)
+import           GHC.Generics
+import           TreeSitter.Cursor as TS
+import           TreeSitter.Language as TS
+import           TreeSitter.Node as TS
+import           TreeSitter.Parser as TS
+import           TreeSitter.Tree as TS
+import           Data.Proxy
+import           Prelude hiding (fail)
 
-data Expression
-      = NumberExpression Number | IdentifierExpression Identifier
-      deriving (Eq, Ord, Show)
+-- Parse source code and produce AST
+parseByteString :: Unmarshal t => Ptr TS.Language -> ByteString -> IO (Either String t)
+parseByteString language bytestring = withParser language $ \ parser -> withParseTree parser bytestring $ \ treePtr ->
+  if treePtr == nullPtr then
+    pure (Left "error: didn't get a root node")
+  else
+    withRootNode treePtr $ \ rootPtr ->
+      withCursor (castPtr rootPtr) $ \ cursor ->
+        runM (runFail (runReader cursor (runReader bytestring buildNode)))
 
-data Number = Number
-      deriving (Eq, Ord, Show)
+-- | Unmarshal is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
+--
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'buildNode' providing that they have a suitable 'Generic' instance.
+class Unmarshal a where
+  buildNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
```

Comment:
Let’s rename `build*` to `unmarshal*` as well 👍 (Whether in this PR or a future one.)